File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode299BullsAndCows.cpp
#include "LeetCode299BullsAndCows.h"

#include <unordered_map>
using namespace std;

string LeetCode299BullsAndCows::getHint(string secret, string guess)
{
    unordered_map<char,int> secretMap{};
    unordered_map<char,int> guessMap{};
    int bulls{};
    int length=secret.length();
    for (int i=0;i<length;i++)
    {
        if (secret[i]==guess[i])
        {
            bulls++;
            continue;
        }
        if (!guessMap.contains(guess[i]))
            guessMap[guess[i]]=0;
        if (!secretMap.contains(secret[i]))
            secretMap[secret[i]]=0;
        secretMap[secret[i]]++;
        guessMap[guess[i]]++;
    }
    int cows{};
    for (auto & pair:secretMap)
    {
        if (!guessMap.contains(pair.first))
            continue;
        cows+=min(pair.second,guessMap[pair.first]);
    }
    return to_string(bulls)+"A"+to_string(cows)+"B";
}

File: LeetCodeProblemsGo/LeetCode299BullsAndCows.go
package main

import "fmt"

func getHint(secret string, guess string) string {
	secretMap := make(map[byte]int)
	guessMap := make(map[byte]int)
	bulls := 0
	length := len(secret)
	for i := 0; i < length; i++ {
		if secret[i] == guess[i] {
			bulls++
			continue
		}
		if _, exists := guessMap[guess[i]]; !exists {
			guessMap[guess[i]] = 0
		}
		guessMap[guess[i]]++
		if _, exists := secretMap[secret[i]]; !exists {
			secretMap[secret[i]] = 0
		}
		secretMap[secret[i]]++
	}
	cows := 0
	for key, value := range secretMap {
		if guessValue, exists := guessMap[key]; exists {
			cows += min(value, guessValue)
		}
	}
	return fmt.Sprintf("%vA%vB", bulls, cows)
}
